stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -490,10 +503,85 @@ impl NakamotoBlockHeader {
+    pub fn verify_threshold_signer(
+        &self,
+        signer_aggregate: &Point,
+        signature: &ThresholdSignature,
+    ) -> bool {
         let message = self.signer_signature_hash().0;
-        schnorr_signature.verify(signer_aggregate, &message)
+        signature.verify(signer_aggregate, &message)
+    }
+
+    /// Verify the block header against the list of signer signatures
+    ///
+    /// Validate against:
+    /// - Any invalid signatures (eg not recoverable or not from a signer)
+    /// - Any duplicate signatures
+    /// - At least the minimum number of signatures (based on total signer weight
+    /// and a 70% threshold)
+    /// - Order of signatures is maintained vs signer set
+    pub fn verify_signer_signatures(&self, reward_set: &RewardSet) -> Result<(), ChainstateError> {
+        let message = self.signer_signature_hash();
+        let Some(signers) = &reward_set.signers else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "No signers in the reward set".to_string(),
+            ));
+        };
+
+        let mut total_weight_signed: u32 = 0;
+        // `last_index` is used to prevent out-of-order signatures
+        let mut last_index = None;
+
+        let total_weight = signers.iter().map(|s| s.weight).sum::<u32>();
+
+        // HashMap of <PublicKey, (Signer, Index)>
+        let signers_by_pk = signers
+            .iter()
+            .enumerate()
+            .map(|(i, signer)| (signer.signing_key, (signer.clone(), i)))
+            .collect::<HashMap<_, _>>();
+
+        for signature in &self.signer_signature {
+            let public_key = Secp256k1PublicKey::recover_to_pubkey(message.bits(), signature)
+                .map_err(|_| {
+                    ChainstateError::InvalidStacksBlock(format!(
+                        "Unable to recover public key from signature {}",
+                        signature.to_hex()
+                    ))
+                })?;
+
+            let (signer, signer_index) = signers_by_pk
+                .get(public_key.to_bytes().as_slice())
+                .ok_or_else(|| {
+                    ChainstateError::InvalidStacksBlock(format!(
+                        "Public key {} not found in the reward set",
+                        public_key.to_hex()
+                    ))
+                })?;
+
+            // Enforce order of signatures
+            match last_index {
+                Some(index) if index >= *signer_index => {
+                    return Err(ChainstateError::InvalidStacksBlock(
+                        "Signatures are out of order".to_string(),
+                    ));
+                }
+                _ => last_index = Some(*signer_index),
+            }
+
+            total_weight_signed += signer.weight;
+        }
+
+        // Calculate 70% of total weight as the threshold
+        let threshold = (total_weight as f64 * 7_f64 / 10_f64).ceil() as u32;
```

Comment:
@kantai Not that it matters anymore since the PoX voting is vestigial, but I believe we intend to use two digits of precision?  So, it's be `total_weight * 70 / 100`.

Reply:
`signer.weight` is essentially the "number of slots" that the signer received. The maximum value for the threshold is `70% of 4000`, and the maximum weight for a given signer is `4000`. That's why these are u32s, but also why my proposed change has no loss of precision: if you set the threshold to be `7 * total / 10`, evaluate the multiplication first, and don't overflow, the resulting threshold could never get more precise (because the signer's weights could only ever increment by "1" anyway.s)

Reply:
Gotcha.

Reply:
I updated the threshold calculation and added a constant to `core/mod.rs`